firefox-ios/Storage/Rust/RustRemoteTabs.swift

```diff
@@ -165,6 +172,207 @@ public class RustRemoteTabs {
+                self.logger.log(err.description,
+                                level: .warning,
+                                category: .tabs)
+                return
+            }
+
+            self.getUnsentCommandsByDeviceId(deviceId: deviceId) { commands in
+                commandQueue.setPendingCommandsSent(deviceId: deviceId, commands: commands) { errors in
+                    if errors.isEmpty {
+                        return
+                    } else {
+                        let errHeader = "Failed to set some pending commands as sent:\n"
+                        let errMessages: String = errors.reduce(errHeader, { result, err in
+                            result.appending("  - \(err.localizedDescription)\n")
+                        })
+                        self.logger.log(errMessages,
+                                        level: .warning,
+                                        category: .tabs)
+                    }
+                }
+            }
+        }
+    }
+
+    private func getUnsentCommandsByDeviceId(deviceId: String, completion: @escaping ([PendingCommand]) -> Void) {
+        guard let commandQueue = self.commandQueue else {
+            let err = TabsApiError.UnexpectedTabsError(reason: "Command queue is not initialized") as MaybeErrorType
+            self.logger.log(err.description,
+                            level: .warning,
+                            category: .tabs)
+            completion([PendingCommand]())
+            return
+        }
+
+        commandQueue.getUnsentCommands { result in
+            switch result {
+            case .success(let commands):
+                let filteredCommands = commands.filter { $0.deviceId == deviceId }
+                completion(filteredCommands)
+            case .failure(let error):
+                self.logger.log("Failed to get unsent commands: \(String(describing: error))",
+                                level: .warning,
+                                category: .tabs)
+                completion([PendingCommand]())
+            }
+        }
+    }
+
+    private func handleCommandQueueErrors(result: Result<Bool, any Error>) {
+        switch result {
+        case .success(let didSucceed):
+            if !didSucceed {
+                self.logger.log("Unexpected error while attempting to update command queue",
+                                level: .warning,
+                                category: .tabs)
+            }
+        case .failure(let error):
+            self.logger.log("Failed to update command queue: \(String(describing: error))",
+                            level: .warning,
+                            category: .tabs)
+        }
+    }
+}
+
+internal class RemoteTabsCommandQueue {
+    var commandStore: RemoteCommandStore?
+
+    init() {}
+
+    func openCommandStore(tabsStore: TabsStore) {
+        self.commandStore = tabsStore.newRemoteCommandStore()
+    }
```

Comment:
Nit / suggestion: It doesn't look like we ever set `commandStore` to `nil`, so I wonder if making it optional, and splitting `openCommandStore` from `init` is just adding a bit of noise, especially since `newRemoteCommandStore()` doesn't do any I/O.

How would you feel about making it a required parameter to the initializer?

```suggestion
    var commandStore: RemoteCommandStore

    init(tabsStore: TabsStore) {
        self.commandStore = tabsStore.newRemoteCommandStore()
    }
```

Reply:
https://github.com/mozilla-mobile/firefox-ios/pull/22799/commits/7d44018db816a8b4494ca1d7b1f12e8113690ce9